packages/jupyter/src/providers.tsx

```diff
@@ -107,3 +108,30 @@ export function ThebeLoaderAndServer({
     </ThebeBundleLoaderProvider>
   );
 }
+
+type OutputsContextType = {
+  allSafe: boolean;
+  outputsId: IdOrKey;
+};
+const OutputsContext = React.createContext<OutputsContextType | null>(null);
+
+export function useOutputsContext() {
+  const context = useContext(OutputsContext);
+  if (context === null) {
+    throw new Error('useOutputsContext must be used within a OutputsContextProvider');
+  }
+  return context;
+}
+export function OutputsContextProvider({
+  allSafe,
+  outputsId,
+  children,
+}: {
+  children: React.ReactNode;
+  allSafe: boolean;
+  outputsId: IdOrKey;
+}) {
+  return (
+    <OutputsContext.Provider value={{ allSafe, outputsId }}>{children}</OutputsContext.Provider>
+  );
+}
```

Comment:
This just allows us to use the `<MyST />` visitor by lifting parent-child state into a context object.

Reply:
yes... this felt a little "Why?!?" to begin with as the context is created in an Outputs renderer and only serves to provide exactly one `outputsId` relating to its parent.

but we're only really using this here to bridge across the interface of the `<MyST />` component, to keep the Output renderer independent.

So (just as discussion here) *if* all renderers received access to the full AST in addition to the node that they were called to render then we potentially would not need this `Context` as an `Output` could look up it's parent, correct? (although we might want it to avoid a lookup) @agoose77 do you know if there has been any discussion on that angle elsewhere?

Reply:
yes... this felt a little "Why?!?" to begin with as the context is created in an Outputs renderer and only serves to provide exactly one `outputsId` relating to its parent.

but we're only really using this here to bridge across the interface of the `<MyST />` component, to keep the Output renderer independent.

So (just as discussion here) *if* all renderers received access to the full AST in addition to the node that they were called to render then we potentially would not need this `Context` as an `Output` could look up it's parent, correct? (although we might want it to avoid a lookup) @agoose77 do you know if there has been any discussion on that angle elsewhere?

Reply:
@stevejpurves yes that's correct. We could drop the `outputID` by making that part of the AST.